### DISCUSSION TITLE
Tests - Where to place?

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#435] [Q&A] [answered]
Куда класть тесты то? FSD это регламентирует?

~~Спойлер: Нет~~

**Comment by azinit** [comment]
```
Есть мнение, что это почти похожая ситуация на тесты
- Можешь класть рядом с файлом, но у тебя разрастется src код (зато все будет "под рукой" и будет сложно где-то хранить общие для сторисов моки и т.п.)
- Можешь класть в отдельную директорию в руте, где у тебя будет обособлена эта ДЕМО-логика от src компонентов (плюс там же можешь хранить нужные утилиты/моки данных и т.п.)
```

https://t.me/feature_sliced/1897

**Reply by azinit** [reply]
```
Я держу рядом с кодом, иначе сложно отслеживать наличие тестов
```

https://t.me/feature_sliced/1908

**Comment by azinit** [comment]
```
Рядом с тестами. А тесты рядом с тестируемыми юнитами.
...
В принципе +, но бывает удобней и отдельно хранить)
А те же моки для сущностей видел, как хранили рядом с сущностью
(особенно если юнит рядом тоже лежат)
```

https://t.me/feature_sliced/5415

**Comment by azinit** [comment]
```
никаких отличий в организации тестов по  FSD или по любой другой архитектуре я не вижу
```

https://t.me/feature_sliced/12315

**Reply by azinit** [reply]
```
Q: Если например есть сущность user и в нем есть файлик ui
Сами тесты по логике я положил в папку tests но по неймингу так и не определился. Сначала думал называть user.test.ts
```

```
A: у тебя логика (допустим одинм файлом) model.ts, почему тогда файл для логики user.test.ts? а не model.test.ts
```

https://t.me/feature_sliced/12318

**Comment by illright** [comment, marked as answer]
**TL;DR: Put them in segments next to the code that they are testing.**

----

There are no restrictions in FSD as to where you can place your tests. The two common configurations are 
1. co-locating tests with the code that they are testing
2. having a separate folder that mirrors the structure of `src` and only contains tests

From the point of view of FSD, the first approach is more preferable, since one of the main benefits of FSD is low coupling, so that changes to one aspect of the project happen in one small subfolder, not one file per folder across all folders. This can be achieved by placing your tests in a `spec`` or `tests/` folder inside each segment. Mocks can be stored in that folder as well.

For example:
* 📂 features
  * 📂 auth
    * 📂 ui
      * LoginButton.tsx
      * 📂 spec
        * LoginButton.spec.tsx
  * 📂 add-note
    * 📂 api
      * useNotes.ts
      * 📂 spec
        * useNotes.spec.ts
        * notes.mock.json
        
Any setup code for tests or global mocks can be placed in its own segment in Shared, `shared/tests`.
